### PR TITLE
[RISCV] Expand mul to shNadd x, (slli x, c) in DAGCombine

### PR DESCRIPTION
**Pull request by preames**
This expansion is directly inspired by the analogous code in the x86 backend for LEA.  shXadd and (this sub-case of) LEA are largely equivalent.

This is an alternative to https://github.com/llvm/llvm-project/pull/87105.

This expansion is also supported via the decomposeMulByConstant callback, but restricted because of interactions with other combines since that code runs before legalization.  As discussed in the other review, my original plan had been to support post legalization expansion through the same interface, but that ended up being more complicated than seems justified.

Instead, lets go ahead and do the general expansion post-legalize.  Other targets use the combine approach, and matching that structure makes it easier for us to adapt ideas from other targets to RISCV.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -13356,10 +13356,56 @@ static SDValue performXORCombine(SDNode *N, SelectionDAG &DAG,
...
-static SDValue performMULCombine(SDNode *N, SelectionDAG &DAG) {
+// Try to expand a scalar multiply to a faster sequence.
+static SDValue expandMul(SDNode *N, SelectionDAG &DAG,
+                         TargetLowering::DAGCombinerInfo &DCI,
+                         const RISCVSubtarget &Subtarget) {
+
   EVT VT = N->getValueType(0);
-  if (!VT.isVector())
+
+  // An imul is usually smaller than the alternative sequence.
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:13366]: imul is an X86 instruction name

File: llvm/test/CodeGen/RISCV/addimm-mulimm.ll [hunk 2]
```diff
@@ -551,8 +551,9 @@ define i64 @add_mul_combine_infinite_loop(i64 %x) {
 ; RV32IMB-NEXT:    sh3add a1, a1, a2
 ; RV32IMB-NEXT:    sh1add a0, a0, a0
 ; RV32IMB-NEXT:    slli a2, a0, 3
-; RV32IMB-NEXT:    addi a0, a2, 2047
-; RV32IMB-NEXT:    addi a0, a0, 1
+; RV32IMB-NEXT:    li a3, 1
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/addimm-mulimm.ll:554]: Is this worse?

Comment (preames) [reply llvm/test/CodeGen/RISCV/addimm-mulimm.ll:554]: a) This test is really checking for a combine loop, and (b) zbs would produce a bset here and zba w/o zbs is mildly odd, and c) the overall better result would probably be (x*3+256)*8 (SH1ADD, ADDI, SHL) so we're not optimal either way.

So yes, but I don't think we care.

**Comment by preames** [comment]
ping?

**Review by topperc — APPROVED** [review]
LGTM

**Comment by dtcxzyw** [comment]
This patch causes llc to generate more instructions than before: https://github.com/dtcxzyw/llvm-codegen-benchmark/commit/2973ba786de69a73026473874d08937415702d63

An example: https://godbolt.org/z/x376qeGEx
```
define i64 @test(i32 signext %x, i32 signext %y) {
  %sub = sub i32 %x, %y
  %ext = zext i32 %sub to i64
  %res = mul nuw nsw i64 %ext, 24
  ret i64 %res
}
```

Before:
```
test:                                   # @test
        subw    a0, a0, a1
        slli.uw a0, a0, 3
        sh1add  a0, a0, a0
        ret
```

After:
```
test:                                   # @test
        subw    a0, a0, a1
        zext.w  a1, a0
        sh1add.uw       a0, a0, a1
        slli    a0, a0, 3
        ret
```

**Comment by preames** [comment]
@dtcxzyw Thanks!  Proposed fix: https://github.com/llvm/llvm-project/pull/88937

**Pull request merged**